runtime/format.cc

```diff
@@ -0,0 +1,330 @@
+//===-- runtime/format.cc ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "format.h"
+#include "../lib/common/format.h"
+#include "../lib/decimal/decimal.h"
+#include <limits>
+
+namespace Fortran::runtime {
+
+FormatControl::FormatControl(Context &context, const char *format,
+    std::size_t formatLength, const MutableModes &modes, int maxHeight)
+  : context_{context}, modes_{modes}, maxHeight_{static_cast<std::uint8_t>(
+                                          maxHeight)},
+    format_{format}, formatLength_{static_cast<int>(formatLength)} {
+  // The additional two items are for the whole string and a
+  // repeated non-parenthesized edit descriptor.
+  if (maxHeight > std::numeric_limits<std::int8_t>::max()) {
+    context_.terminator.Crash(
+        "internal Fortran runtime error: maxHeight %d", maxHeight);
+  }
+  stack_[0].start = offset_;
+  stack_[0].remaining = Iteration::unlimited;  // 13.4(8)
+}
+
+int FormatControl::GetMaxParenthesisNesting(
+    Terminator &terminator, const char *format, std::size_t formatLength) {
+  using Validator = common::FormatValidator<char>;
+  Validator::Reporter reporter{[&](const common::FormatMessage &message) {
+    terminator.Crash(message.text, message.arg);
+    return false;  // crashes on error above
+  }};
+  Validator validator{format, formatLength, reporter};
+  validator.Check();
+  return validator.maxNesting();
+}
+
+void FormatControl::SkipBlanks() {
+  while (offset_ < formatLength_ && format_[offset_] == ' ') {
+    ++offset_;
+  }
+}
+
+char FormatControl::GetNextChar() {
+  if (offset_ >= formatLength_) {
+    context_.terminator.Crash("FORMAT missing at least one ')'");
```

Comment:
Might this message look strange if it ever gets generated?  (Or is that the expectation, that it's a fallback that should probably never be generated?)

Reply:
It's what will come out if format control crashes out of the end of the string, which it should do only if there's an insufficiency of `)` characters to close out the format and any nested parenthesized groups.